options/logger_registry.go

```diff
@@ -0,0 +1,112 @@
+package options
+
+import (
+	"encoding/json"
+	"sync"
+
+	"github.com/mongodb/grip/send"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+// LoggerRegistry is an interface that stores reusable logger factories.
+type LoggerRegistry interface {
+	Register(string, LoggerFactory)
+	Check(string) bool
+	Names() []string
+	Resolve(string) (LoggerFactory, bool)
+}
+
+type basicLoggerRegistry struct {
+	mu        sync.RWMutex
+	factories map[string]LoggerFactory
+}
+
+func NewBasicLoggerRegistry() LoggerRegistry {
+	return &basicLoggerRegistry{
+		factories: map[string]LoggerFactory{},
+	}
+}
+
+func (r *basicLoggerRegistry) Register(name string, factory LoggerFactory) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.factories[name] = factory
+}
+
+func (r *basicLoggerRegistry) Check(name string) bool {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	_, ok := r.factories[name]
+	return ok
+}
+
+func (r *basicLoggerRegistry) Names() []string {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	names := []string{}
+	for name := range r.factories {
+		names = append(names, name)
+	}
+
+	return names
+}
+
+func (r *basicLoggerRegistry) Resolve(name string) (LoggerFactory, bool) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	factory, ok := r.factories[name]
+	return factory, ok
+}
+
+// LogConfigFormat describes the format of the log configuration.
+type LogConfigFormat string
+
+const (
+	LogConfigFormatBSON LogConfigFormat = "BSON"
+	LogConfigFormatJSON LogConfigFormat = "JSON"
+)
+
+func (f LogConfigFormat) Validate() error {
+	switch f {
+	case LogConfigFormatBSON, LogConfigFormatJSON:
+	default:
+		return errors.New("unknown log config format")
+	}
+
+	return nil
+}
+
+// Unmarshal unmarshals the data into out based on the log config format.
+func (f LogConfigFormat) Unmarshal(data []byte, out interface{}) error {
```

Comment:
I think this should become private. 

also I think out can be of LoggerProducer type